### PR TITLE
Add a DpProposer to enhance collective_plan execution speed

### PR DESCRIPTION
**Pull request by tiankongdeguiji**
The collective_plan will encounter an NCCL timeout if the number of EmbeddingBagConfig exceeds 500, as noted in #2282 , To address this issue, we propose using a Proposer with dynamic programming to improve the execution speed of the collective plan.

In the case of `test_multiple_ebc_plan.py` mentioned in #2282, the collective plan with the DpProposer took only 11.86 seconds, whereas the original version took 904.90 seconds.

### CODE REVIEW

**Comment by tiankongdeguiji** [comment]
Hi, @henrylhtsang  @IvanKobzarev  @joshuadeng  @PaulZhang12  can you review this?

**Comment by henrylhtsang** [comment]
that sounds interesting, @sarckk do you want to take a look?

File: torchrec/distributed/planner/proposers.py [hunk 1]
```diff
@@ -279,6 +280,147 @@ def feedback(
...
+    def propose(self) -> Optional[List[ShardingOption]]:
+        """Propose a sharding plan."""
+        if not self._inited:
+            proposal_index = [0] * len(self._sharding_options_by_fqn)
+            return [
+                self._sharding_options_by_fqn[fqn][index]
+                for fqn, index in zip(
+                    self._sharding_options_by_fqn.keys(), proposal_index
+                )
+            ]
```
Comment (shinoss) [inline torchrec/distributed/planner/proposers.py]: nit 
```suggestion
        if not self._inited:
            return [
                sharding_options[0] for sharding_options in self._sharding_options_by_fqn.values()
            ]
```

File: torchrec/distributed/planner/planners.py [hunk 2]
```diff
@@ -188,9 +189,7 @@ def __init__(
             )
         else:
             self._proposers = [
-                GridSearchProposer(),
-                GreedyProposer(),
-                GreedyProposer(use_depth=False),
+                DpProposer(),
```
Comment (shinoss) [inline torchrec/distributed/planner/planners.py]: Let's keep default proposers the same for backward compatibility

```suggestion
                GridSearchProposer(),
                GreedyProposer(),
                GreedyProposer(use_depth=False),
```

File: torchrec/distributed/planner/proposers.py [hunk 3]
```diff
@@ -279,6 +280,147 @@ def feedback(
...
+                    hbm_i = int(hbm_by_fqn[0][j])
+                    dp[0][j][hbm_i] = (perf_by_fqn[0][j], hbm_by_fqn[0][j])
+
+            for i in range(1, M):
+                for j in range(N):
+                    for c in range(K):
+                        if c + hbm_by_fqn[i][j] >= K:
+                            continue
+                        for nj in range(N):
+                            prev_perf, perv_hbm = dp[i - 1][nj][c]
```
Comment (shinoss) [inline torchrec/distributed/planner/proposers.py]:
```suggestion
                            prev_perf, prev_hbm = dp[i - 1][nj][c]
```

File: torchrec/distributed/planner/proposers.py [hunk 4]
```diff
@@ -279,6 +280,147 @@ def feedback(
...
+                            if prev_perf < float("inf"):
+                                new_hbm = perv_hbm + hbm_by_fqn[i][j]
```
Comment (shinoss) [inline torchrec/distributed/planner/proposers.py]:
```suggestion
                                new_hbm = prev_hbm + hbm_by_fqn[i][j]
```

File: torchrec/distributed/planner/proposers.py [hunk 5]
```diff
@@ -279,6 +280,147 @@ def feedback(
         self._proposal_index += 1
 
 
+def _bytes_to_float_bin(num_bytes: Union[float, int], bin_size: float) -> float:
+    return float(num_bytes) / bin_size
+
+
+class DpProposer(Proposer):
```
Comment (shinoss) [inline torchrec/distributed/planner/proposers.py]: nit: rename to `DynamicProgrammingProposer` as `Dp` in torchrec usually refers to Data Parallel

```suggestion
class DynamicProgrammingProposer(Proposer):
```

File: torchrec/distributed/planner/proposers.py [hunk 6]
```diff
@@ -279,6 +280,147 @@ def feedback(
...
+        """Feedback last proposed plan."""
+        if not self._inited:
+            self._inited = True
+            M = len(self._sharding_options_by_fqn)
+            N = max([len(x) for x in self._sharding_options_by_fqn.values()])
+
+            assert storage_constraint is not None
+            hbm_total = sum([x.storage.hbm for x in storage_constraint.devices])
+            K = self._hbm_bins_per_device * len(storage_constraint.devices)
+            bin_size = float(hbm_total) / K
```
Comment (shinoss) [inline torchrec/distributed/planner/proposers.py:388]: let's be defensive and guard against `bin_size = 0` which will result in division by 0 in `_bytes_to_float_bin`

**Review by sarckk — CHANGES_REQUESTED** [review]
Hi @tiankongdeguiji thanks for this PR. I've been thinking about this for a while, but I don't fully understand how this DP is set up, and the purpose of multiple HBM bins. 

To my understanding, this proposer tries to explore the full space of possibilities (num_embedding_tables x num_sharding_variant x num_possible_partitions) assuming total HBM size is uniformly distributed across K equally-sized HBM bins. M * N * K represents the total search space of sharding type + partitioning these shards. Is this correct? If so, how do you tune the number of HBM bins per device? 

Could you provide some high-level description of the DP algorithm, as a docstring for this proposer? Thanks

File: torchrec/distributed/planner/planners.py (hunk 2 above)
Comment (tiankongdeguiji) [reply torchrec/distributed/planner/planners.py]: In all of our test scenarios, the DynamicProgrammingProposer is faster than both the GreedyProposer and GridSearchProposer. Additionally, the performance of the sharding plans generated by the DynamicProgrammingProposer is better. Which scenario might encounter issues?

**Comment by tiankongdeguiji** [comment]
> Hi @tiankongdeguiji thanks for this PR. I've been thinking about this for a while, but I don't fully understand how this DP is set up, and the purpose of multiple HBM bins.
> 
> To my understanding, this proposer tries to explore the full space of possibilities (num_embedding_tables x num_sharding_variant x num_possible_partitions) assuming total HBM size is uniformly distributed across K equally-sized HBM bins. M * N * K represents the total search space of sharding type + partitioning these shards. Is this correct? If so, how do you tune the number of HBM bins per device?
> 
> Could you provide some high-level description of the DP algorithm, as a docstring for this proposer? Thanks

I provide a high-level overview of the DP algorithm in docstring. 

The number of HBM bins affects both the plan counts and plan precision. We will generate plans with a total HBM constraint ranging from 0 to K in increments of the bin width, and keep the best perf plan in each bin. In most cases, we do not need to adjust the number of HBM bins per device, 100 bins is generally sufficient. However, if the number of tables becomes several thousand, we may need additional bins.

**Comment by tiankongdeguiji** [comment]
In the `test_multiple_ebc_plan.py` case mentioned in #2282 , the collective plan with the DynamicProgrammingProposer has been further optimized to 1.48 seconds, while the first commit version of the DynamicProgrammingProposer took 11.86 seconds, and TorchRec now takes 904.90 seconds. @sarckk, could you take another look?

**Comment by sarckk** [comment]
@tiankongdeguiji thanks for the PR!

**Pull request closed without merging**